src/ape/api/projects.py

```diff
@@ -1,667 +1,134 @@
-import os.path
-import re
-from collections.abc import Iterator, Sequence
+from functools import cached_property
 from pathlib import Path
-from typing import TYPE_CHECKING, Any, Optional, Union
+from typing import Optional
 
-from ethpm_types import Checksum, Compiler, ContractType, PackageManifest, Source
-from ethpm_types.source import Content
-from packaging.version import InvalidVersion, Version
-from pydantic import AnyUrl, ValidationError
+from pydantic import field_validator
 
-from ape.exceptions import ProjectError
-from ape.logging import logger
-from ape.utils import (
-    BaseInterfaceModel,
-    ExtraAttributesMixin,
-    ExtraModelAttributes,
-    abstractmethod,
-    cached_property,
-    create_tempdir,
-    get_all_files_in_directory,
-    get_relative_path,
-    log_instead_of_fail,
-)
+from ape.api.config import ApeConfig
+from ape.utils import BaseInterfaceModel, abstractmethod
 
-if TYPE_CHECKING:
-    from ape.contracts import ContractContainer
 
-
-class ProjectAPI(BaseInterfaceModel):
+class DependencyAPI(BaseInterfaceModel):
```

Comment:
wonder if this having an "is immutable?" boolean might be helpful to basically describe whether it should be "trusted" as a dependency not to change, or always check to see if there is some integrity checksum that has been modified to drive re-acquiring the dependency (and recompiling it with the context of the old version to gain a performance boost from the old project's cache)

For example, github releases essentially shouldn't change (caveat here that they actually can, but typically don't), ipfs links will not change, http links definitely can change, so can git branches etc.

Reply:
Or maybe a property "needs resync?" that returns whether the integrity sum is still valid, or (in the case of IPFS) just is always true

For git repos, can store the commit hash of that commit for the branch/tag

Reply:
> For example, github releases essentially shouldn't change 

They can if you are using `ref: main` or something.

Re: resyncing, I think we can make an API out of this somehow but I would consider it extra, I think we could do it though. I think it is extra because when in doubt, `ape pm install --force` again, as a workaround. Also it is seems a bit complicated but maybe we can make a ticket for this.

Reply:
> > For example, github releases essentially shouldn't change
> 
> They can if you are using `ref: main` or something.

Yeah that was sort of my point. Anyways, my second suggestion is actually better if you can just use the commit hash as an integrity hash (e.g. you can quickly see if there are any changes to try and recompile with by checking to see if the upstream ref mismatches the stored commit hash you have for that ref)

> Re: resyncing, I think we can make an API out of this somehow but I would consider it extra, I think we could do it though. I think it is extra because when in doubt, `ape pm install --force` again, as a workaround. Also it is seems a bit complicated but maybe we can make a ticket for this.

I would think this would live under `DependencyAPI`, but unsure of where this resync logic lives now with the refactor

Reply:
> I would think this would live under DependencyAPI, but unsure of where this resync logic lives now with the refactor

There is no re-sync feature as of now, as far as I know

Reply:
But yes, we would add it to DependencyAPI and then DependencyManager would also have a method probably

Reply:
> But yes, we would add it to DependencyAPI and then DependencyManager would also have a method probably. And when requesting a dependency that the manager detects needs a refresh, we could have that happen.
> 
> Still though - this is a bit extra! Does pip even do that with `github` based installs?

Usually if you install a git dependency, it just always fetches and recompiles it. Unsure if any further caching is going on, but that could also just be the way we handle it (although I think it might be more annoying for us to always go and fetch dependency sources)

Reply:
There is definitely a difference in how pip installs packages, and how we get the files to compile dependencies

Should the dependency sources just live on disk until you manually refresh? that could also be fine, worth some thought

Reply:
> Should the dependency sources just live on disk until you manually refresh? that could also be fine, worth some thought

This is how it works now, and it is pretty simple i think so i like that.